doc/architecture/006-Stateful-deployments.md

```diff
@@ -0,0 +1,26 @@
+# Stateful deployments
+
+## Summary
+
+This document motivates why NixOps 4 supports stateful deployments, and explains why deployment state is optional and first-class.
+
+## Context
+
+NixOps 1 and 2 are stateful deployment tools, in the sense that it maintains a database with information about each deployment.
+
+This has been a point of criticism, because it makes puts a burden on the user to make sure that the database is used correctly. NixOps 2 has mitigated this by supporting automatic synchronization of the database, but this feature is not enabled by default, and was "bolted on" to the existing application architecture.
+
+Meanwhile, `nixos-rebuild switch --target-host <host>` is a stateless deployment tool. It simply consumes the information passed to it, and does not require any information to be stored and reused between invocations.
+
+However, in cloud deployments, preserving state is often necessary. For instance, when a cloud instance is created, this creates information, such as the SSH public host key, that is necessary to connect to the instance, and may not be stored in the cloud provider's API for us to retrieve later. Even then, retrieving this information may be slow.
+
```

Comment:
This should mention deletions.

Deletions are different in stateless vs stateful.
stateless: delete by invoking a certain command before removing it from the expressions
stateful: delete from expressions, let NixOps figure out what to delete.